collectors/modality-probe-debug-collector/src/lib.rs

```diff
@@ -0,0 +1,1169 @@
+
+use fenced_ring_buffer::async_reader::{FencedReader, Snapper};
+use fenced_ring_buffer::{FencedRingBuffer, SeqNum, WholeEntry};
+use modality_probe::log::LogEntry;
+use modality_probe::{
+    DynamicHistory, EventId, LogicalClock, ModalityProbe, OrdClock, ProbeEpoch, ProbeId, ProbeTicks,
+};
+use modality_probe_collector_common::{
+    add_log_report_to_entries, json::write_log_entries, Report, ReportLogEntry, SessionId,
+};
+
+/// Either a u32 or u64, depending on the target architecture
+#[derive(Debug, PartialEq, Copy, Clone, Eq, Hash)]
+pub enum Word {
+    U32(u32),
+    U64(u64),
+}
+
+impl Word {
+    fn size(&self) -> u8 {
+        match self {
+            Self::U32(_) => 4,
+            Self::U64(_) => 8,
+        }
+    }
+
+    #[cfg(test)]
+    fn unwrap_32(&self) -> u32 {
+        match self {
+            Self::U32(n) => *n,
+            Self::U64(_) => panic!("Tried to unwrap u32 from U64 Word"),
+        }
+    }
+}
+
+impl Into<usize> for Word {
+    fn into(self) -> usize {
+        match self {
+            Self::U32(n) => n as usize,
+            Self::U64(n) => n as usize,
+        }
+    }
+}
+
+impl Add<u64> for Word {
+    type Output = Word;
+
+    // Note: u64s added to 32 bit words must not be greater than u32::MAX
+    fn add(self, rhs: u64) -> Word {
+        match self {
+            Word::U32(n) => Word::U32(n + u32::try_from(rhs).unwrap()),
+            Word::U64(n) => Word::U64(n + rhs),
+        }
+    }
+}
+
+// Offset of pointer to DynamicHistory in ModalityProbe struct, which is located in modality-probe/src/lib.rs
+fn history_ptr_offset() -> u64 {
+    offset_of!(ModalityProbe => history).get_byte_offset() as u64
+}
+
+// Address offsets of each needed field of the DynamicHistory struct, which is located in modality-probe/src/history.rs
+fn overwrite_priority_offset() -> u64 {
+    offset_of!(DynamicHistory => overwrite_priority).get_byte_offset() as u64
+}
+
+fn probe_id_offset() -> u64 {
+    offset_of!(DynamicHistory => probe_id).get_byte_offset() as u64
+}
+
+fn write_seqn_high_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => write_seqn: SeqNum => high)
+        .get_byte_offset() as u64
+}
+
+fn write_seqn_low_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => write_seqn: SeqNum => low)
+        .get_byte_offset() as u64
+}
+
+fn overwrite_seqn_high_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => overwrite_seqn: SeqNum => high)
+        .get_byte_offset() as u64
+}
+
+fn overwrite_seqn_low_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => overwrite_seqn: SeqNum => low)
+        .get_byte_offset() as u64
+}
+
+fn log_storage_addr_offset() -> u64 {
+    offset_of!(DynamicHistory => log: FencedRingBuffer<LogEntry> => storage).get_byte_offset()
+        as u64
+}
+
+fn log_storage_cap_offset(n_word_bytes: u8) -> u64 {
+    // Assume storage addr is a u32
```

Comment:
What's the basis for this assumption?  Is it one we can/should check at runtime?

Reply:
Sorry, this comment was outdated, `n_word_bytes` is 4 or 8 depending on the target word size.